### ISSUE TITLE
remove "with sharing" from core classes that don't need it (AFTER security review is complete)

### ISSUE THREAD

**Issue by mscholtz** [SalesforceFoundation/Cumulus#1479]
There are many classes in NPSP that have "with sharing" declaration that IMHO don't warrant it, and I humbly suggest that you review where this is really needed (VF controllers, etc)  and where it's not (most trigger code IMHO).  

This has caused me several problems with writing code that will work w/ NPSP3.  In every case this is code that worked fine w/ NPSP2, and we made expected adjustments, but hard-to-debug permissions errors have gotten in the way (hard to debug because occurring on managed NPSP code).

One situation was posted on the PUB here (although discussion petered out because I didn't have it diagnosed quite properly): https://powerofus.force.com/0D58000001PuE6G

Another that just arose was this:

* client has Sharing Settings on opp set to Read Only
* we have a custom Before Insert trigger on opp that resets the owner
* an integration (IATS Brickwork) inserts opps from donation forms via a Sites guest user account
* worked fine in NPSP2
* in NPSP3, was throwing a "insufficient access rights on cross-reference ID" error
* through copious testing, was able to figure out that the error was happening here: https://github.com/SalesforceFoundation/Cumulus/blob/2d8fc54af04102c9aa84b68380eedd5587d6b95e/src/classes/TDTM_TriggerHandler.cls#L149
* my understanding is that after the before trigger changes the owner of the opp, the Guest User account no longer has permissions to edit it.  when your trigger fires and calls this class (declared as "with sharing") a DML of some sort that touches opps is no longer allowed and throws this err
* If I comment out the resetting of Owner, or if I change the Sharing Settings to Read/Write, the error doesn't occur

My understanding of the TDTM_TriggerHandler class is limited, but it doesn't seem like the sort of thing that needs to be "with sharing".  Most triggers should just run when the conditions are met, regardless of which user put the data in.

Furthermore, it leads to the kind of errors and difficult debugging that I've been dealing with today.  I think you don't want your users to have to assign additional permissions to user accounts simply to prevent core NPSP  triggers from throwing errors.

Thanks! 
@davidhabib @ceiroa

**Comment by mscholtz** [comment]
A few supporting documents to look at:

* http://www.tgerm.com/2011/03/trigger-insufficient-access-cross.html (sharing declarations still matter when called from triggers)
* http://salesforce.stackexchange.com/questions/16121/sfdc-understanding-with-sharing-without-sharing-unspecified-sharing-classes (first answer by sfdcfox) - this gibes with my understanding: with sharing is essential for most anything that has a UI (VF pages, etc.), but without sharing is often what you want for business process triggers that should occur without UI regardless of which user created the starting conditions

**Comment by kbromer** [comment]
@ceiroa this dovetails nicely with one of our actions items for HEDAP - lets definitely discuss-

**Comment by ceiroa** [comment]
I think you are correct, @mscholtz. I was under the wrong impression that any class called from a trigger would run in system mode, even if it was labeled "with sharing", due to a cascading effect. It seems like the cascading effect applies to classes calling other classes (as described in the [post you reference from sfdcfox](http://salesforce.stackexchange.com/questions/16121/sfdc-understanding-with-sharing-without-sharing-unspecified-sharing-classes)), but not to [triggers calling classes](http://www.tgerm.com/2011/03/trigger-insufficient-access-cross.html).

Changing TDTM_TriggerHandler to be "without sharing" is an easy enough change, and it seems like the only one that's necessary. Every other class can still be "with sharing", and the sharing model will apply if they are called from controllers and such, which is nice from a reusability perspective.

I'll do some more research and testing, and hopefully include this change shortly.

**Comment by mscholtz** [comment]
Awesome, thank you for looking into this.  The other thing you could do is
just leave classes undeclared.  I think that works fine most of the time.
If it gets called from a trigger, it's effectively "without sharing", while
if it gets called from a VF controller that's decleared "with sharing",
it'll inherit that.

If a class has a declaration, my understanding is that's the mode it runs
in, regardless of what calls it.  OTOH, if no declaration exists, it
respects the mode of what called it.



On Fri, Mar 27, 2015 at 12:49 PM, Carlos <notifications@github.com> wrote:

> I think you are correct, @mscholtz <https://github.com/mscholtz>. I was
> under the wrong impression that any class called from a trigger would run
> in system mode, even if it was labeled "with sharing", due to a cascading
> effect. It seems like the cascading effect applies to classes calling other
> classes (as described in the post you reference from sfdcfox
> <http://salesforce.stackexchange.com/questions/16121/sfdc-understanding-with-sharing-without-sharing-unspecified-sharing-classes>),
> but not to triggers calling classes
> <http://www.tgerm.com/2011/03/trigger-insufficient-access-cross.html>.
>
> Changing TDTM_TriggerHandler to be "without sharing" is an easy enough
> change, and it seems like the only one that's necessary. Every other class
> can still be "with sharing", and the sharing model will apply if they are
> called from controllers and such, which is nice from a reusability
> perspective.
>
> I'll do some more research and testing, and hopefully include this change
> shortly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SalesforceFoundation/Cumulus/issues/1479#issuecomment-87069776>
> .
>



-- 

Matthew Scholtz
Percolator Consulting
(206) 289-0816 x3

http://percolatorconsulting.com/

**Comment by ceiroa** [comment]
@mscholtz - I'm starting to think the option you propose might be the only available one. I believe I initially misunderstood the StackOverflow post (hard to follow with all the numbering...), and cascading does not apply to classes calling classes either, if "with sharing" or "without sharing" is used. Only when those keywords are not used can you count on the cascading effect. 

From the [official documentation](https://www.salesforce.com/us/developer/docs/apexcode/Content/apex_classes_keywords_sharing.htm):

* The sharing setting of the class where the method is defined is applied, not of the class where the method is called. For example, if a method is defined in a class declared with with sharing is called by a class declared with without sharing, the method will execute with sharing rules enforced.
* If a class isn’t declared as either with or without sharing, the current sharing rules remain in effect. This means that the class doesn’t enforce sharing rules except if it acquires sharing rules from another class. For example, if the class is called by another class that has sharing enforced, then sharing is enforced for the called class.

**Comment by njjc** [comment]
That's great to know, and different from what we thought when discussing this the other day. Let's remove all "with sharing" references, and leave any "without sharing" references in place.

**Comment by ceiroa** [comment]
Well, not all. We should not remove them from controllers.

**Comment by kbromer** [comment]
Just be warned, any place that doesn't explicitly declare it will require an explanation for SecRev

**Comment by barbarachristensen** [comment]
Found another instance where "with sharing" seems to be causing an issue: Merging an existing Contact into a Household Account, with "add and merge households". In an org where sharing on Account is set by role hierarchy, users aren't able to fully merge Household Accounts they don't own., as they can't delete the Account from which they are merging. It renames it "Anonymous Household" w/ the Address record and the Primary Contact still associated, even though the Contact has been merged into a new HH. 
thx

**Comment by judisohn** [comment]
Barbara, I'm not sure if this is the same issue, as even a standard merge wouldn't work between accounts a user doesn't own unless the user has 'Modify All' either at the object or system permission level of their profile/permission set. The ability to delete records one does not own can't be controlled from Sharing Rules.

**Comment by mscholtz** [comment]
@judisohn while it's true that sharing rules don't affect deletion permissions, I do believe that "with sharing" does affect it.  If it's a trigger doing the deletion (and in this case it is an NPSP trigger) then if it's "with sharing" it will apply the end user's permissions (leading to the error), but if it's not "with sharing" it'll run in system mode and avoid the err.

To be clear, this err happens for our end user whether they use the Merge HH button, or just manually change the HH lookup field and let the triggers do the merge.  If we start w/ two accounts that the user already owns, the err does not occur.

**Comment by judisohn** [comment]
Ah, now that makes sense. Thanks.

**Comment by force2b** [comment]
Related issue posted in the Hub: https://powerofus.force.com/0D580000023095l

**Comment by njjc** [comment]
So I have removed "with sharing" from all TDTM classes. I'm wondering if there are any other good candidates. @davidhabib, what do you think? All the merge related classes you've been working on is probably a good idea.

**Comment by davidhabib** [comment]
or if you don't have perms, you should be merging!  i think it would break security if you could merge records you don't have perms on.  I think we should start with all the TDTM classes, and get that out there, but hold off on any additional until we have a use case we understand.

**Comment by njjc** [comment]
The problem is having perms on the account of contacts you're merging and do have permissions to merge. I don't think you could search for records you don't have permissions for in the first place.

**Comment by mscholtz** [comment]
Thanks, guys.  Can someone post here when this fix gets pushed out?  Or is there some way for me to tell that on my own?

Oh wait, did this get reverted?  @njjc?

**Comment by mscholtz** [comment]
@davidhabib understand the caution, but I think it's probably safe to leave any helper classes undeclared, including those that are involved in something like merging.  That's because if the merge is launched from a VF page, and the controller is declared "with sharing", then the helper classes will inherit that and will impose sharing.  But OTOH if you decide to reuse some of the code in the helper class from a trigger (i.e. you set up some sort of auto-merge), you don't have to worry about unexpectedly hitting errors due to a sharing limitation.

**Comment by njjc** [comment]
It did get reverted. We're currently going through SF security review and are worried that this change will be a red flag. Security review will finish within the next month (hopefully within the next two weeks) and we'll merge it in after.

**Comment by mrbelvedere** [comment]
Included in beta release version 3.46 (Beta 1) and higher

**Comment by mrbelvedere** [comment]
Included in production release version 3.46

**Comment by force2b** [comment]
Interesting, I ran into this issue again in our Sandbox while building a REST service published through a Force.com site (i.e., authentication not required for the REST service). Version 3.54 is installed in the Sandbox. The REST Service was failing when updating the Contact object.

The error message returned was "List has no rows for assignment to SObject". A clip from the debug log is included below. I was able to work-around the issue by creating a Criteria-Based Sharing Rule on the Contact object that shared all Contacts with the Site's "Guest" User (I created a Public Group with that Guest User as the only member). That is definitely not a secure work-around, but it works for now.

Debug Log:
```
00:54:14.435 (435519567)|ENTERING_MANAGED_PKG|npsp
00:54:14.475 (475420316)|SAVEPOINT_SET|[76]|SavepointValue0
00:54:14.487 (487592845)|SOQL_EXECUTE_BEGIN|[50]|Aggregations:0|SELECT COUNT(Id) rowCount FROM Trigger_Handler__c
00:54:14.490 (490243712)|SOQL_EXECUTE_END|[50]|Rows:1
00:54:14.491 (491384236)|SOQL_EXECUTE_BEGIN|[70]|Aggregations:0|SELECT npsp__Asynchronous__c, npsp__Class__c, npsp__Load_Order__c, npsp__Object__c, npsp__Active__c, npsp__Trigger_Action__c FROM Trigger_Handler__c WHERE (Active__c = TRUE AND Object__c = :tmpVar1 AND Trigger_Action__c INCLUDES (:tmpVar2)) ORDER BY Load_Order__c ASC NULLS FIRST
00:54:14.497 (497309034)|SOQL_EXECUTE_END|[70]|Rows:4
00:54:14.527 (527302918)|SOQL_EXECUTE_BEGIN|[95]|Aggregations:0|SELECT Id, npe01__One2OneContact__c, npe01__SYSTEM_AccountType__c FROM Account WHERE Id = :tmpVar1
00:54:14.557 (557175107)|SOQL_EXECUTE_END|[95]|Rows:0
00:54:14.574 (574069607)|SOQL_EXECUTE_BEGIN|[77]|Aggregations:0|SELECT Id, npe01__SYSTEM_AccountType__c FROM Account WHERE Id = :tmpVar1
00:54:14.580 (580616291)|SOQL_EXECUTE_END|[77]|Rows:0
00:54:14.581 (581942605)|SAVEPOINT_ROLLBACK|[123]|SavepointValue0
00:54:14.601 (601522953)|DML_BEGIN|[182]|Op:Insert|Type:npsp__Error__c|Rows:1
00:54:14.659 (659561252)|DML_END|[182]
00:54:14.667 (667884564)|SOQL_EXECUTE_BEGIN|[172]|Aggregations:0|select Id, Name from CollaborationGroup where id = :chatterGroupId
00:54:14.671 (671590821)|SOQL_EXECUTE_END|[172]|Rows:0
```

I'll add that no log entry was ever created even though the log implies that it was successfully inserted.

**Comment by mscholtz** [comment]
What is the status of this, folks?  It was closed a while ago, but looks like there are still a lot of non-controller classes marked as "with sharing".

**Comment by kbromer** [comment]
Do you have something specific to point to? A behavior that isn't working as expected?  The ones mentioned above in the thread should be resolved AFAIK.  Some of them just need proper sharing setup in the org.

**Comment by mscholtz** [comment]
The error that @barbarachristensen describes in her comment above from April 13 is still happening.  And there's @force2b's issue from just above.

Not sure what you mean by "proper sharing setup".

**Comment by force2b** [comment]
Only that triggers shouldn't fail with that error since they're supposed to run in the System Context. Because the issue was resolved by sharing the Contacts with the guest user, this indicates that somewhere along the path in the TDTM Contact triggers, code is running that explicitly uses "With Sharing". I'm unable to pinpoint it down to a specific trigger because the logs are for managed packages are limited.

Looking through the repository, many of the TDTM classes are still declared "with sharing". I can't point to a specific class only because the logs don't expose that, but I'm pretty confident that the root cause are the TDTM related classes still using "with sharing" (or possibly a services class called by the TDTM handler). There's a note from 4/30 that says all "with sharing" references were removed from the TDTM classes. Did that change get reverted?

**Comment by kbromer** [comment]
Well, for @force2b I'd argue that you should explicitly share records to the guest user if you want to do that, otherwise anybody with your URL can potentially manipulate those records, you've basically stripped away a pretty big line of defense.  That said, I'm fairly certain all TDTM classes have had that with sharing removed, so its tough to say where he could be running into the issue, I'm not sure what the sites user is doing in what he described. 

For @barbarachristensen I'd say that's WAD - since you don't want a user in an org with a role hierarchy to bypass the hierarchy you've setup to do a merge.

**Comment by mscholtz** [comment]
https://github.com/SalesforceFoundation/Cumulus/blob/e0525f8a93d1fb7246940d2fa35c37c503149a42/src/classes/TDTM_TriggerHandler.cls
https://github.com/SalesforceFoundation/Cumulus/blob/e0525f8a93d1fb7246940d2fa35c37c503149a42/src/classes/TDTM_ObjectDataGateway.cls
few more, as well as classes like this:
https://github.com/SalesforceFoundation/Cumulus/blob/e0525f8a93d1fb7246940d2fa35c37c503149a42/src/classes/ADDR_Validator.cls

@kbromer I see your point re Barbara's issue.  I think we still have a client that's seeing problems when HH triggers run when they do operations like taking someone out of a HH (by blanking out the Account field on a contact) but we need to verify that.  Will get back to you on that.

**Comment by force2b** [comment]
I'm using a private API key in the REST request to provide some level authentication. Still, generally speaking apex code should be able to update any record in the system even if the currently logged in User doesn't have rights to that record. That's a fundamental part of using without sharing, even going beyond the Guest user. 

Taking this issue beyond my REST service, through the use of Sharing Rules, we can prevent users from modifying records through standard functionality (i.e., a page layout) by making the record **or field* read-only or even private. A visualforce page that is exposed to the user can permit specific actions and then perform CRUD on that object as per business requirements. That should always work, however in this case the NPSP triggers are blocking updates to the Contact object even though the bit of code that is committing the update has the permissions to do so.

Here are a few TDTM classes using "with sharing":
* https://github.com/SalesforceFoundation/Cumulus/blob/e0525f8a93d1fb7246940d2fa35c37c503149a42/src/classes/TDTM_ObjectDataGateway.cls
* https://github.com/SalesforceFoundation/Cumulus/blob/e0525f8a93d1fb7246940d2fa35c37c503149a42/src/classes/TDTM_TriggerActionHelper.cls
* https://github.com/SalesforceFoundation/Cumulus/blob/e0525f8a93d1fb7246940d2fa35c37c503149a42/src/classes/TDTM_TriggerHandler.cls

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x800000001hAoAAI" target="_blank">W-008722</a>

**Comment by force2b** [comment]
Thanks for re-opening and looking into the classes still declared as With Sharing.

In this particular use case, I'm not attempting to go around assigning a user. The REST services are part of a website integration that uses GET and POST methods to query for events and allow someone to register for an event, which creates a Contact within Salesforce.

Anyway, it interesting that you both disagree on the use of "without sharing". I've employed tight sharing and even tighter FLS in scenarios where I want to completely protect an object or fields (since "sharing" here applies to both sharing rules, object CRUD and FLS) from a user making any modification directly - whether that be through editing on a page layout or via an import (Users can import Accounts/Contacts in some orgs, though not typically NPO's). In doing so, I may build a Visualforce page or Asynchronous process (such as a callout to retrieve additional data) that will update that object and those fields. In that use-case, the related Apex runs as the Users' context (i.e., they don't have all permissions to the record or field), but the apex will still be able to update the record because it was declared as "Without Sharing".  Of course all that said, my default for all Apex Classes is always "with sharing", or sometimes undeclared where I need it to inherit the calling classes mode. "Without Sharing" has a very specific use, and should never be applied liberally.

**Comment by kbromer** [comment]
Ah, gotcha. 

This will make for a good conversation over a beer at some point :beers:
